source/extensions/filters/network/zookeeper_proxy/decoder.cc

```diff
@@ -476,41 +476,101 @@ void DecoderImpl::skipStrings(Buffer::Instance& data, uint64_t& offset) {
+                                                   DecodeType dtype) {
+  uint32_t data_len = data.length();
+  uint32_t zk_filter_buffer_len = zk_filter_buffer.length();
+
+  if (zk_filter_buffer_len > 0) {
+    if (zk_filter_buffer_len + data_len > INT_LENGTH) {
+      // Prepend ZooKeeper filter buffer to current buffer.
+      data.prepend(zk_filter_buffer);
+      decodeAndBufferHelper(data, zk_filter_buffer, dtype);
+      // Drain the prepended ZooKeeper filter buffer.
+      data.drain(zk_filter_buffer_len);
+    } else {
+      zk_filter_buffer.add(data);
+    }
+  } else if (zk_filter_buffer_len == 0) {
+    decodeAndBufferHelper(data, zk_filter_buffer, dtype);
+  }
+  return Network::FilterStatus::Continue;
+}
+
+void DecoderImpl::decodeAndBufferHelper(Buffer::Instance& data, Buffer::OwnedImpl& zk_filter_buffer,
+                                        DecodeType dtype) {
+  uint32_t data_len = data.length();
+  uint32_t zk_filter_buffer_len = zk_filter_buffer.length();
   uint64_t offset = 0;
-  int32_t len = 0;
+  uint32_t len = 0;
+  // Boolean to check whether there is at least one full packet in the buffer.
+  bool has_full_packets = false;
 
-  while (offset < data.length()) {
+  while (offset < data_len) {
     try {
       // Peek packet length.
       len = helper_.peekInt32(data, offset);
-      ensureMinLength(len, INT_LENGTH + XID_LENGTH);
+      if (dtype == DecodeType::READ) {
+        ensureMinLength(len, XID_LENGTH + INT_LENGTH);
+      } else if (dtype == DecodeType::WRITE) {
+        ensureMinLength(len, XID_LENGTH + ZXID_LENGTH + INT_LENGTH);
+      }
       ensureMaxLength(len);
       offset += len;
+      if (offset <= data_len) {
+        has_full_packets = true;
+      }
     } catch (const EnvoyException& e) {
       ENVOY_LOG(debug, "zookeeper_proxy: decoding exception {}", e.what());
       callbacks_.onDecodeError();
-      return Network::FilterStatus::Continue;
+      return;
     }
   }
 
-  if (offset == data.length()) {
+  if (offset == data_len) {
     decode(data, dtype);
-    return Network::FilterStatus::Continue;
-  }
 
-  ENVOY_LOG(trace,
-            "zookeeper_proxy: waiting for entire packets in the buffer, current buffer "
-            "is {} bytes",
-            data.length());
-  return Network::FilterStatus::StopIteration;
+    if (zk_filter_buffer_len > 0) {
+      // Drain the ZooKeeper filter buffer.
+      zk_filter_buffer.drain(zk_filter_buffer_len);
+    }
+  } else if (offset > data_len) {
+    std::string temp_data;
+
+    if (has_full_packets) {
+      offset -= INT_LENGTH + len;
+      // Decode full packets.
+      temp_data.reserve(offset);
+      temp_data.resize(offset);
+      data.copyOut(0, offset, temp_data.data());
+      Buffer::OwnedImpl full_packets;
+      full_packets.add(temp_data.data(), temp_data.length());
+      decode(full_packets, dtype);
+
+      // Drain the ZooKeeper filter buffer.
+      zk_filter_buffer.drain(zk_filter_buffer_len);
```

Comment:
again why we drain here? we drain in the caller?

Reply:
This is because after prepending the ZK filter buffer, the decoder is able to decode all full packets above. Since we used the data in the ZK filter buffer already, we need to drain the ZK filter buffer here and add the remaining partial data from the network filter to the ZK filter buffer, which will be prepended to the network filter buffer next time.

Let me update the comment.